compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -827,10 +820,167 @@ impl InitMask {
+        self.start = end_of_chunk;
+
+        ret
+    }
+}
+
+/// Returns the index of the first bit in `start..end` (end-exclusive) that is equal to is_init.
+fn find_bit(init_mask: &InitMask, start: Size, end: Size, is_init: bool) -> Option<Size> {
+    fn find_bit_fast(init_mask: &InitMask, start: Size, end: Size, is_init: bool) -> Option<Size> {
+        fn search_block(
+            bits: Block,
+            block: usize,
+            start_bit: usize,
+            is_init: bool,
+        ) -> Option<Size> {
+            // invert bits so we're always looking for the first set bit
+            let bits = if is_init { bits } else { !bits };
+            // mask off unused start bits
+            let bits = bits & (!0 << start_bit);
+            // find set bit, if any
+            if bits == 0 {
+                None
+            } else {
+                let bit = bits.trailing_zeros();
+                Some(size_from_bit_index(block, bit))
+            }
+        }
+
+        if start >= end {
+            return None;
+        }
+
+        let (start_block, start_bit) = bit_index(start);
+        let (end_block, end_bit) = bit_index(end);
+
+        // handle first block: need to skip `start_bit` bits
+        if let Some(i) =
+            search_block(init_mask.blocks[start_block], start_block, start_bit, is_init)
+        {
+            if i < end {
+                return Some(i);
+            } else {
+                // if the range is less than a block, we may find a matching bit after `end`
+                return None;
+            }
+        }
+
+        let one_block_past_the_end = if end_bit > 0 {
+            // if `end_bit` > 0, then the range overlaps `end_block`
+            end_block + 1
+        } else {
+            end_block
+        };
+
+        // handle remaining blocks
+        if start_block < one_block_past_the_end {
+            for (&bits, block) in init_mask.blocks[start_block + 1..one_block_past_the_end]
+                .iter()
+                .zip(start_block + 1..)
+            {
+                if let Some(i) = search_block(bits, block, 0, is_init) {
+                    if i < end {
+                        return Some(i);
+                    } else {
+                        // if this is the last block, we may find a matching bit after `end`
+                        return None;
+                    }
+                }
+            }
+        }
+
+        None
+    }
+
+    #[cfg_attr(not(debug_assertions), allow(dead_code))]
+    fn find_bit_slow(init_mask: &InitMask, start: Size, end: Size, is_init: bool) -> Option<Size> {
+        (start..end).find(|&i| init_mask.get(i) == is_init)
+    }
+
+    let result = find_bit_fast(init_mask, start, end, is_init);
+
+    debug_assert_eq!(
+        result,
+        find_bit_slow(init_mask, start, end, is_init),
+        "optimized implementation of find_bit is wrong for start={:?} end={:?} is_init={} init_mask={:#?}",
```

Comment:
Tests + stage 2 build passed with this changed to `assert_eq`, so I'm pretty sure it's correct

Reply:
That's good! However, I think the code could be made easier to read by adding a few comments. In particular I am missing the big picture here -- there are a lot of small steps and I have no idea why they are all needed. I am not entirely getting this `one_block_past_the_end` business.

Also, helper functions like `search_block` should have comments explaining what exactly they do.